pkg/storage/stores/shipper/bloomshipper/block_downloader.go

```diff
@@ -0,0 +1,224 @@
+		queue:              downloadingQueue,
+		blockClient:        blockClient,
+		activeUsersService: activeUsersService,
+		limits:             limits,
+	}
+
+	for i := 0; i < config.BlocksDownloadingQueue.WorkersCount; i++ {
+		go b.serveDownloadingTasks(fmt.Sprintf("worker-%d", i))
+	}
+	return b
+}
+
+type BlockDownloadingTask struct {
+	ctx context.Context
+	// Request is the original request
+	block BlockRef
+	// ErrCh is a send-only channel to write an error to
+	ErrCh chan<- error
+	// ResultsCh is a send-only channel to return the path to the directed where the block is extracted
+	ResultsCh chan<- BlockDownloadingResult
+}
+
+func NewBlockDownloadingTask(ctx context.Context, block BlockRef, resCh chan<- BlockDownloadingResult, errCh chan<- error) *BlockDownloadingTask {
+	return &BlockDownloadingTask{
+		ctx:       ctx,
+		block:     block,
+		ErrCh:     errCh,
+		ResultsCh: resCh,
+	}
+}
+
+func (d *blockDownloader) serveDownloadingTasks(workerID string) {
+	logger := log.With(d.logger, "worker", workerID)
+	level.Debug(logger).Log("msg", "starting worker", "worker", workerID)
+
+	d.queue.RegisterConsumerConnection(workerID)
+	defer d.queue.UnregisterConsumerConnection(workerID)
+
+	idx := queue.StartIndexWithLocalQueue
+
+	for {
+		item, newIdx, err := d.queue.Dequeue(context.Background(), idx, workerID)
+		if err != nil {
+			if !errors.Is(err, queue.ErrStopped) {
+				level.Error(logger).Log("msg", "failed to dequeue task", "err", err)
+				continue
+			}
+			level.Info(logger).Log("msg", "stopping worker")
+			return
+		}
+		task, ok := item.(*BlockDownloadingTask)
+		if !ok {
+			level.Error(logger).Log("msg", "failed to cast to BlockDownloadingTask", "item", fmt.Sprintf("%+v", item), "type", fmt.Sprintf("%T", item))
+			continue
+		}
+
+		idx = newIdx
+		blockPath := task.block.BlockPath
+		level.Debug(logger).Log("msg", "start downloading the block", "block", blockPath)
+		block, err := d.blockClient.GetBlock(task.ctx, task.block)
+		if err != nil {
+			level.Error(logger).Log("msg", "error downloading the block", "block", blockPath, "err", err)
+			task.ErrCh <- fmt.Errorf("error downloading the block %s : %w", blockPath, err)
+			continue
+		}
+		directory, err := d.extractBlock(&block, time.Now())
+		if err != nil {
+			level.Error(logger).Log("msg", "error extracting the block", "block", blockPath, "err", err)
+			task.ErrCh <- fmt.Errorf("error extracting the block %s : %w", blockPath, err)
+			continue
+		}
+		task.ResultsCh <- BlockDownloadingResult{block: task.block, extractedBlockPath: directory}
+	}
+}
+
+type BlockDownloadingResult struct {
+	extractedBlockPath string
+	block              BlockRef
+}
+
+func (d *blockDownloader) downloadBlocks(ctx context.Context, tenantID string, references []BlockRef) (chan blockWithQuerier, chan error) {
+	d.activeUsersService.UpdateUserTimestamp(tenantID, time.Now())
+	errCh := make(chan error)
+	blocksCh := make(chan blockWithQuerier, len(references))
+	go func() {
```

Comment:
I'm questioning whether we want to spawn a goroutine for each `downloadBlocks` call. Enqueueing is already non-blocking.

I think you can avoid that by returning actual results by the `downloadBlocks()` function (instead of the channels). That would remove the `for { select {} }` in the `ForEachBlock()` function of the shipper.

Reply:
I want to return the channels because some blocks might be already in the cache, so we can start processing those immediately but for the rest, we will have to download and extract them, which might take some time ... 
wdyt?

Reply:
fixed